### ISSUE TITLE
NegativeBinomial grad with k = 0 failures

### ISSUE THREAD

**Issue by staples6** [TuringLang/Turing.jl#680]
https://github.com/TuringLang/Turing.jl/blob/0d5cf235e2224e61e3876b92ad57b901e1342eff/src/core/ad.jl#L187

If I understand the parameterization of the NegativeBinomial being used, I believe  _nbinomlogpdf_grad_1 would need to range from 0:k since there can be 0 failures before the rth success? Is that correct?

I believe that is the cause for the following error:

```julia
using Turing

@model nbexample(y, r) = begin
        p ~ Beta(98, 2)
        y ~ NegativeBinomial(r, p)
end

chn = sample(nbexample(0, 10), NUTS(1000, 100, .7))

ERROR: ArgumentError: reducing over an empty collection is not allowed
```

**Comment by mohamed82008** [comment]
I don't think this is the correct fix. When k is 0, this is just the probability of having `r` successes, which is `p^r`. The partial of `log(p^r)` wrt `r` is `log(p)`. The second partial derivative nicely handles the `k = 0` case.

Note that in the above code, I used `k + r - i` not `k + r - 1 - i` so I went from `i = 1:k` as opposed to `i = 0:k-1` which is what you would get from the choose operator. So the above code is correct AFAICT. It is also validated against finite difference. The fix is simple, we can do `mapreduce(i -> 1 / (k + r - i), +, 1:k, init = zero(p)) + log(p)` to handle the empty collection case.
